### PR TITLE
Adding cluster cli feature

### PR DESCRIPTION
**Pull request by alaydshah**
Cluster commands:

```
❯ fedml cluster -h
Usage: fedml cluster [OPTIONS] COMMAND [ARGS]...

  Manage clusters on the MLOps platform.

Options:
  -h, --help  Show this message and exit.

Commands:
  kill      Kill (tear down) clusters from the MLOps platform.
  killall   Kill (tear down) ALL clusters from the MLOps platform.
  list      List clusters from the MLOps platform.
  start     Start clusters from the MLOps platform.
  startall  Start ALL clusters from the MLOps platform.
  status    Status of clusters from the MLOps platform.
  stop      Stop clusters from the MLOps platform.
  stopall   Stop ALL clusters from the MLOps platform.
```

### CODE REVIEW

File: python/fedml/api/modules/cluster.py
Comment (chaoyanghe) [inline python/fedml/api/modules/cluster.py:1]: let's don't use "click" related APIs in fedml.api

File: python/fedml/computing/scheduler/comm_utils/security_utils.py
Comment (chaoyanghe) [inline python/fedml/computing/scheduler/comm_utils/security_utils.py:1]: what's the machnamism of save_api_key and get_api_key?

File: python/fedml/cli/modules/cluster.py
Comment (chaoyanghe) [inline python/fedml/cli/modules/cluster.py:1]: I think we need to support "stop/start/shutdown/status/list". Stop and stop are not implemented yet.

File: python/fedml/api/modules/cluster.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+
+def print_clusters(cluster_list_obj):
+    click.echo("Found the following matching clusters.")
+    cluster_list_table = PrettyTable(['Cluster Name', 'Cluster ID', 'Status'])
+
+    for cluster in cluster_list_obj.cluster_list:
+        cluster_list_table.add_row([cluster.cluster_name, cluster.cluster_id, cluster.status])
+
+    print(cluster_list_table)
```
Comment (chaoyanghe) [inline python/fedml/api/modules/cluster.py:54]: print_clusters is not used by CLI, please use "_single_leading_underscore" style: _print_clusters() as "internal use" indicator. FYI:
https://peps.python.org/pep-0008/#descriptive-naming-styles

File: python/fedml/computing/scheduler/scheduler_entry/launch_manager.py [hunk 2]
```diff
@@ -782,7 +774,8 @@ def __init__(self, job_yaml_file, should_use_default_workspace=False):
         self.base_dir = os.path.dirname(job_yaml_file)
         self.using_easy_mode = True
         self.executable_interpreter = "bash"
-        self.executable_file_folder = self.job_config_dict.get("workspace", None) \
+        self.executable_file_folder = os.path.normpath(
```
Comment (fedml-alex) [inline python/fedml/computing/scheduler/scheduler_entry/launch_manager.py:777]: If workspace is a absolute path, the self.executable_file_folder will be not correct.

File: python/fedml/computing/scheduler/scheduler_entry/cluster_manager.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        if response.status_code != 200:
+            print(
+                f"Cluster list with response.status_code = {response.status_code}, response.content: {response.content}")
+            pass
+        else:
+            resp_data = response.json()
+            code = resp_data.get("code", None)
+            data = resp_data.get("data", None)
+            if code is None or data is None or code == "FAILURE":
+                print(f"List job with response.status_code = {response.status_code}, "
```
Comment (fedml-alex) [inline python/fedml/computing/scheduler/scheduler_entry/cluster_manager.py:117]: The text 'List job' should be 'Cluster list'.

File: python/fedml/cli/modules/cluster.py
Comment (alaydshah) [reply python/fedml/cli/modules/cluster.py:1]: Yes, waiting on backend endpoints for it. I will take care of these and other remaining commands in another PR

File: python/fedml/api/modules/cluster.py (hunk 1 above)
Comment (alaydshah) [reply python/fedml/api/modules/cluster.py:54]: Thank you for the catch. Will update

File: python/fedml/computing/scheduler/comm_utils/security_utils.py
Comment (alaydshah) [reply python/fedml/computing/scheduler/comm_utils/security_utils.py:1]: It saves in the `launch_secret` file under `$HOME/fedml-client/fedml/data/secret` directory, and reads back when needed from the same file

image

File: python/fedml/computing/scheduler/scheduler_entry/cluster_manager.py (hunk 3 above)
Comment (alaydshah) [reply python/fedml/computing/scheduler/scheduler_entry/cluster_manager.py:117]: Good catch. Will fix. Thanks

File: python/fedml/computing/scheduler/scheduler_entry/launch_manager.py (hunk 2 above)
Comment (alaydshah) [reply python/fedml/computing/scheduler/scheduler_entry/launch_manager.py:777]: I think we have trade off to make:

1. With my modified implementation, workspace needs to always be relative path to `job.yaml`. The upside is, `fedml` commands can be ran from outside the directory and would still work. For eg, if `job.yaml` is in example folder (`example/job.yaml`), you need NOT require to run `fedml` from within example  directory and can run from anywhere as long as you give correct path to `job.yaml` i.e. `fedml launch /path/to/job.yaml`  would work.

2. With current implementation, workspace can be either relative or absolute path. But then, one has to always run `fedml` command from within the directory where `job.yaml` is present i.e. `fedml launch /path/to/job.yaml ` won't work. one would have to `cd` to `/path/to/job.yaml` and then execute `fedml launch job.yaml`

What do we think is more convenient to users?

File: python/fedml/api/modules/cluster.py
Comment (alaydshah) [reply python/fedml/api/modules/cluster.py:1]: Addressed for cluster APIs

**Comment by chaoyanghe** [comment]
image
@alaydshah asce has updated new APIs, please add and further PR.

**Comment by alaydshah** [comment]
> image @alaydshah asce has updated new APIs, please add and further PR.

Hi,
Have done bit more refactoring and also added `start` and `stop` commands. PR is ready to re-review

**Pull request merged**